lib/Mir.ml

```diff
@@ -13,76 +13,59 @@ open Core_kernel
 
 type litType = Int | Real | Str [@@deriving sexp, hash]
 
-type operator =
-  | Plus
-  | Minus
-  | Times
-  | Divide
-  | Modulo
-  | Or
-  | And
-  | Equals
-  | NEquals
-  | Less
-  | Leq
-  | Greater
-  | Geq
-[@@deriving sexp, hash]
-
-type transformation = expr Ast.transformation
-
-and index =
+and 'e index =
   | All
-  | Single of expr
+  | Single of 'e
   (*
-  | MatrixSingle of expr
+  | MatrixSingle of 'e
  *)
-  | Upfrom of expr
-  | Downfrom of expr
-  | Between of expr * expr
-  | MultiIndex of expr
+  | Upfrom of 'e
+  | Downfrom of 'e
+  | Between of 'e * 'e
+  | MultiIndex of 'e
 
 (** XXX
 *)
-and expr =
+and 'e expr =
   | Var of string
   | Lit of litType * string
-  | FunApp of string * expr list
-  | BinOp of expr * operator * expr
-  | TernaryIf of expr * expr * expr
-  | Indexed of expr * index list
+  | FunApp of string * 'e list
+  | TernaryIf of 'e * 'e * 'e
+  | Indexed of 'e * 'e index list
 [@@deriving sexp, hash]
 
-type adtype = Ast.autodifftype [@@deriving sexp, hash]
-type sizedtype = expr Ast.sizedtype [@@deriving sexp, hash]
 type unsizedtype = Ast.unsizedtype [@@deriving sexp, hash]
+type autodifftype = Ast.autodifftype [@@deriving sexp, hash]
+
+let no_loc = {Ast.filename= ""; line_num= 0; col_num= 0; included_from= None}
+let no_span = {Ast.begin_loc= no_loc; end_loc= no_loc}
 
 (* This directive silences some spurious warnings from ppx_deriving *)
 [@@@ocaml.warning "-A"]
 
-type fun_arg_decl = (adtype * string * unsizedtype) list
+type fun_arg_decl = (autodifftype * string * unsizedtype) list
 
-and 's statement =
-  | Assignment of expr * expr
-  | TargetPE of expr
-  | NRFunApp of string * expr list
-  | Check of string * expr list
+and ('e, 's) statement =
+  | Assignment of 'e * 'e
+  | TargetPE of 'e
+  | NRFunApp of string * 'e list
+  | Check of string * 'e list
   | Break
   | Continue
-  | Return of expr option
+  | Return of 'e option
   | Skip
-  | IfElse of expr * 's * 's option
-  | While of expr * 's
+  | IfElse of 'e * 's * 's option
+  | While of 'e * 's
   (* XXX Collapse with For? *)
-  | For of {loopvar: expr; lower: expr; upper: expr; body: 's}
+  | For of {loopvar: 'e; lower: 'e; upper: 'e; body: 's}
   (* A Block for now corresponds tightly with a C++ block:
      variables declared within it have local scope and are garbage collected
      when the block ends.*)
   | Block of 's list
   (* An SList does not share any of Block's semantics - it is just multiple
      (ordered!) statements*)
   | SList of 's list
-  | Decl of {decl_adtype: adtype; decl_id: string; decl_type: unsizedtype}
+  | Decl of {decl_adtype: autodifftype; decl_id: string; decl_type: unsizedtype}
```

Comment:
Still not 100% sure whether we'll need sizes on these types. My sense now is that the sizes here should be optional. For function inlining (which can be important in combination with other optimizations and dependence analysis), it is important to be able to have decls without sizes. However, for optimization purposes, it could be good to also have optional sizes. Do we want to add those sizes later?

@bob-carpenter , @seantalts , what do you think?

Reply:
These are just for local variables, which aren't declared with sizes from what I remember, right? So any sizes we'd be adding would be ones the optimizer figured out. Wouldn't that go in the metadata somewhere then? Either in the 'e expr metadata or whatever graphs and other structures y'all are building up in the dataflow analysis and optimizer passes?

Reply:
Local variables are declared with a size as far as I'm aware, just not with a transformation. I guess we need to enforce the size at runtime if it's declared. That would suggest optional sizes might be the way to go. For purposes of code gen/optimization, there's even `auto` in C++ if we want to be really lazy and we don't want to enforce any type constraints.

Reply:
Oh right it's just the function parameters that don't have sizes! We can make a new type type that has optional sizes and change the Decl to have one of those.